File: idle-game.go
package main

import (
    "fmt"
    "time"
)

// Player represents the player in the game.
type Player struct {
    ClickPower int
    Gold       int
}

// Upgrade represents an upgrade available for purchase.
type Upgrade struct {
    Name        string
    ClickPower  int
    Cost        int
    Description string
}

func main() {
    player := Player{
        ClickPower: 1,
        Gold:       0,
    }

    upgrades := []Upgrade{
        {"Click Power +1", 1, 10, "Increase your click power by 1"},
        {"Click Power +5", 5, 50, "Increase your click power by 5"},
        {"Click Power +10", 10, 100, "Increase your click power by 10"},
    }

    fmt.Println("Welcome to the Incremental Clicker Game!")
    fmt.Println("Earn gold by clicking. Use gold to purchase upgrades.")

    go autoCollectGold(&player)

    for {
        fmt.Println("Gold:", player.Gold)
        fmt.Println("Click Power:", player.ClickPower)
        fmt.Println("Upgrades Available:")
        for i, upgrade := range upgrades {
            fmt.Printf("%d. %s - Cost: %d - %s\n", i+1, upgrade.Name, upgrade.Cost, upgrade.Description)
        }

        fmt.Println("Enter the number of the upgrade you want to purchase (or 0 to exit):")
        var choice int
        fmt.Scanln(&choice)

        if choice == 0 {
            fmt.Println("Thanks for playing!")
            break
        }

        if choice > 0 && choice <= len(upgrades) {
            upgrade := upgrades[choice-1]
            if player.Gold >= upgrade.Cost {
                player.Gold -= upgrade.Cost
                player.ClickPower += upgrade.ClickPower
                fmt.Println("Upgrade purchased:", upgrade.Name)
            } else {
                fmt.Println("Not enough gold to purchase upgrade.")
            }
        } else {
            fmt.Println("Invalid choice. Please enter a valid upgrade number.")
        }
    }
}

func autoCollectGold(player *Player) {
    for {
        player.Gold += player.ClickPower
        time.Sleep(1 * time.Second)
    }
}

File: idle-game.pas
program ResourceClicker;

uses
  SysUtils, CRT;

const
  INITIAL_RESOURCE_RATE = 1; // Initial resource generation rate per second
  UPGRADE_COST = 10; // Cost of upgrading resource generation rate
  UPGRADE_MULTIPLIER = 2; // Multiplier for resource generation rate upon upgrade

var
  resources: Integer; // Current amount of resources
  resourceRate: Integer; // Current resource generation rate per second
  lastUpdateTime: LongInt; // Time of the last update

// Function to update the game state
procedure UpdateGameState;
var
  currentTime: LongInt;
  elapsedTime: Integer;
begin
  currentTime := Round(1000 * (Now - EncodeDate(2024, 1, 1))); // Current time in milliseconds since game start
  elapsedTime := currentTime - lastUpdateTime; // Time elapsed since last update

  resources := resources + resourceRate * (elapsedTime div 1000); // Update resources based on resource generation rate

  lastUpdateTime := currentTime; // Update last update time
end;

// Function to display the game UI
procedure DisplayUI;
begin
  ClrScr;
  WriteLn('Resource Clicker');
  WriteLn('----------------');
  WriteLn('Resources: ', resources);
  WriteLn('Resource Generation Rate: ', resourceRate, ' per second');
  WriteLn;
  WriteLn('Options:');
  WriteLn('1. Click to Collect Resources');
  WriteLn('2. Upgrade Resource Generation Rate (Cost: ', UPGRADE_COST, ' resources)');
  WriteLn('3. Quit');
end;

// Function to handle player input
procedure HandleInput;
var
  choice: Integer;
begin
  Write('Enter your choice: ');
  ReadLn(choice);
  
  case choice of
    1: resources := resources + 1; // Click to collect resources
    2: begin // Upgrade resource generation rate
         if resources >= UPGRADE_COST then
         begin
           resources := resources - UPGRADE_COST;
           resourceRate := resourceRate * UPGRADE_MULTIPLIER;
           UPGRADE_COST := Round(UPGRADE_COST * 1.5); // Increase upgrade cost for next upgrade
         end
         else
           WriteLn('Insufficient resources to upgrade!');
       end;
    3: Halt; // Quit the game
    else
      WriteLn('Invalid choice. Please try again.');
  end;
end;

// Main game loop
procedure GameLoop;
var
  currentTime: LongInt;
begin
  lastUpdateTime := Round(1000 * (Now - EncodeDate(2024, 1, 1))); // Initialize last update time
  while True do
  begin
    currentTime := Round(1000 * (Now - EncodeDate(2024, 1, 1))); // Current time in milliseconds since game start
    if currentTime - lastUpdateTime >= 1000 then // Update game state every second
    begin
      UpdateGameState;
      DisplayUI;
    end;
    HandleInput;
  end;
end;

// Main program
begin
  resources := 0;
  resourceRate := INITIAL_RESOURCE_RATE;
  GameLoop;
end.
